Fix random failure in setup_zdup

Yvesyingyi opened the pull request:
This issue may be caused by quick transition between tests. In bootloader, a return key is sent to enter grub screen; in setup_zdup, screen is being immidiately evaluated. The os may not have reacted to this point, thus the wrong screen is being evaluated, causing an error. Adding sleep(2) will ensure the os can finish the reaction.

- Related ticket: https://progress.opensuse.org/issues/50804
- Needles: Not needed
- Verification run: http://10.67.17.131/tests/27#step/setup_zdup/1

asmorodskyi commented:
1. Highly unlikely that putting 2 seconds extra slip before `wait_boot` which waiting for appropriate screen for 4 minutes can solve some issues 
2. at least o3 job mentioned in the ticket not fail anymore with error which mentioned here https://openqa.opensuse.org/tests/924078

Yvesyingyi commented:
I speculate this is a random failure. Sometimes the screen failed to respond before `wait_boot`, in which case a wrong needle is matched, and for 4 minutes `wait_boot` is waiting for the wrong screen. Given the randomness, waiting for 2 extra seconds should be enough to reduce the chance of failure.

asmorodskyi commented:
Something wrong with the test flow if we trying to match same needle two times in a row ( 'inst-bootmenu' ) and this is root cause of the problem . inserting sleeps is just workaround of main issue

asmorodskyi commented:
also taking into account randomness of the issue , one verification run does not prove anything . You need to do much more to see if this sleep really solving something

Yvesyingyi commented:
> Something wrong with the test flow if we trying to match the same needle two times in a row

I think the core cause of the problem is at `opensusebasetest.pm:501`. 

``` perl
sub wait_grub_to_boot_on_local_disk {
    # assuming the cursor is on 'installation' by default and 'boot from
    # harddisk' is above
    send_key_until_needlematch 'inst-bootmenu-boot-harddisk', 'up';
    boot_local_disk;
```

It didn't take into account the possibility of slow response by the system. Changing this may affect too many tests in my opinion

Yvesyingyi commented:
in `wait_boot` -> `wait_grub`:
``` perl
    elsif (match_has_tag('inst-bootmenu')) {
        $self->wait_grub_to_boot_on_local_disk;
    }
```

asmorodskyi commented:
another thing about this test case that you booting from iso to just select that you want to boot from HDD . Maybe it will be easier / faster / safer just to boot from HDD ? 
also you can simply drop bootloader module and it should work , because wait_boot will handle all

Yvesyingyi commented:
> Maybe it will be easier / faster / safer just to boot from HDD?
> also you can simply drop bootloader module and it should work because wait_boot will handle all

Yes it is. Added param BOOT_HDD_IMAGE=1, dropped sleep(2), deleted load_boot_tests(), worked fine
http://10.67.17.131/tests/30#

okurz commented:
Please start your commit message with capital letter and fix the typo s/faliure/failure/

Yvesyingyi commented:
Thank you Oliver! 

I just noticed that at where I removed `load_boot_tests`:
``` perl
elsif (installzdupstep_is_applicable()) {
    load_boot_tests();
    load_zdup_tests();
}
```
the condition `installzdupstep_is_applicable`:
``` perl
sub installzdupstep_is_applicable {
    return !get_var("NOINSTALL") && !get_var("RESCUECD") && get_var("ZDUP");
}
```
You're right, it means only if `NOINSTALL` is *not* specified, `load_boot_tests` will run, and I better not remove `load_boot_tests`. Maybe I should change it back

Yvesyingyi commented:
Couldn't reproduce the issue locally, not sure my changes will work, still WIP

Now I'm trying to use `openqa-clone-custom-git-refspec` and see if I can reproduce it on `openqa.opensuse.org`

Yvesyingyi commented:
@okurz I think I've got an effective verification run for the `sleep` method. All of the tests below try the problematic part of `setup_zdup` up to 10 times:

http://10.67.18.100/tests/66#next_previous
Before `sleep 2` fix: test 66 to 70, 4 failed, 1 passed
After `sleep 2` fix: test 71 to 75, all passed

IMO, the fix works, and sleeping for 2 seconds has a minimal impact on efficiency.
